### ISSUE TITLE
Scalar Retiming

### ISSUE THREAD

**Issue by yaqiz01** [raghup17/plasticine#15]
In case of splitting, one UnitPipeline can go through multiple CUs. In this case we need to buffer the scalar values into FIFO in case of delay mismatch. On the producer side, we've already have the connection to spread out scalar outputs to different words on a bus. However, on FIFO read side, we would need the ability to coalescing 16 words read from FIFO back to the first lane. 
@raghup17 @dkoeplin

**Comment by raghup17** [comment]
Ah yes.
One way we can support this in hardware is as follows:

Producer side: spread out scalar values from lane 0 to multiple words using muxes at the output bus.
Currently, it seems like only r0 and r1 in lane0 can contain scalar values.

Consumer side: 
  - write to FIFO: Couple of points here:

       - We currently hard wire input buses to scratchpads. This means that scalar values on each input bus goes into its own scratchpad which is configured as a FIFO.
       - Scratchpads can currently be configured as either fully chained (banks =1), or fully banked (banks = 16). This is an issue if we want to write <16 elements in parallel. One dumb solution is to configure the FIFO as a banked FIFO, and write all 16 values anyway. Given that only the first 2 words will contain valid data, only the first 2 FIFO banks will contain scalar data. We end up writing garbage, and also reducing the effective FIFO size, but I think it works.

- read from FIFO:
      - Currently, registers 12 - 15 are configured to be scalar registers in the empty stage. We can add a path from the first two words from each FIFO read port (4 x 2 words) to scalarInXbar. A 2:1 Mux in front of scalarInXbar can choose whether its inputs come directly from the input bus, or from the FIFO read port. This would require 4 MUXes.

In summary:
Hardware support: 
  - MUXes at producer's output bus, word0 and word1.
  - MUXes in front of each scalarInXbar input, choosing between (word0, word1) of input bus and (word0, word1) of FIFO read port.

Software support:
   - MUX config to spread out data.
   - Config to use scratchpads as parallel FIFOs, configure enq and deq signals with correct delay.
   - MUX config to choose scalarInXbar inputs.

Let me know if this works, and we can talk about how to express this in PISA.
